### PR TITLE
Process Data Models

### PR DESCRIPTION
**Pull request by Ibizugbe**
In this PR, I have completed the current milestone of the project and kept to the requirements of the project.
- no linter errors
- the project is professionally documented.
- the proper git-flow practice has been used.
- model classes have been set up for all entities as shown in the ER diagram.
- the associations have been set up between models.
- a method that returns the 3 most recent posts for a given user has been created.
- a method that updates the posts counter for a user has been created.
- a method that returns the 5 most recent comments for a given post has been created.
- a method that updates the comments counter for a post has been created.
- a method that updates the likes counter for a post has been created.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts
+  has_many :comments
+  has_many :likes
```
Comment (rloterh) [inline app/models/user.rb]: - [ ] As per the given _ERD_, the _User_ **association** should have _author_id_ referenced as **foreign key** in _Post_, _Comment,_ and _Like_ models. Kindly ensure to improve this aspect, as can be seen below:
```ruby
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] Foreign key already declared in the _User_ model, and hence no need to declare it again in a referenced model. Simply do something like:
```ruby
# app/models/post.rb
belongs_to :author, class_name: 'User'

File: app/models/comment.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (rloterh) [inline app/models/comment.rb]: - [ ] Foreign key already declared in the _User_ model, and hence no need to declare it again in a referenced model. Simply do something like:
```ruby
# app/models/comment.rb
belongs_to :author, class_name: 'User'

File: app/models/like.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class Like < ApplicationRecord
+  belongs_to :post, class_name: 'Post', foreign_key: 'post_id'
+  belongs_to :author, class_name: 'User', foreign_key: 'author_id'
```
Comment (rloterh) [inline app/models/like.rb]: - [ ] Foreign key already declared in the _User_ model, and hence no need to declare it again in a referenced model. Simply do something like:
```ruby
# app/models/like.rb
belongs_to :author, class_name: 'User'

File: app/models/like.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+  after_save :update_like_counter
+
+  def update_like_counter
+    post.increment!(:LikesCounter)
+  end
```
Comment (rloterh) [inline app/models/like.rb:11]: - [ ] OPTIONAL
Following best practices `update_like_counter` should be marked as private after implementing the `after_save` method.

File: app/models/comment.rb [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+
+  after_save :update_comment_counter
+
+  def update_comment_counter
+    post.increment!(:CommentsCounter)
+  end
```
Comment (rloterh) [inline app/models/comment.rb:11]: - [ ] OPTIONAL
Following best practices `update_comment_counter` should be marked as private after implementing the `after_save` method.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Abraham, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- Nice job implementing the `after_save` method after data is saved

## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Kindly check the comments marked **[OPTIONAL]** under the review._

**You can also consider:**
- I think you should change the header title _Authors_ to _Author_ in your ReadMe, as this project is worked on by a sole contributor. 
- MIT license link in your ReadMe is broken, kindly endeavor to fix it.
Implementing the above change(s) will make your ReadMe/ repository stand out even better 😉

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/user.rb (hunk 1 above)
Comment (Ibizugbe) [reply app/models/user.rb]: Thanks a lot for correcting these errors, I will make corrections immediately.

File: app/models/post.rb (hunk 2 above)
Comment (Ibizugbe) [reply app/models/post.rb]: Alright, I will make changes immediately.

File: app/models/comment.rb (hunk 3 above)
Comment (Ibizugbe) [reply app/models/comment.rb]: OK, I will do well to make corrections immediately.

File: app/models/like.rb (hunk 4 above)
Comment (Ibizugbe) [reply app/models/like.rb]: Thanks for pointing this out, I will make corrections now.

**Comment by Ibizugbe** [comment]
Thanks, @rloterh  for taking out time to review my project, I will implement these corrections immediately.

File: app/models/comment.rb (hunk 6 above)
Comment (descholar-ceo) [reply app/models/comment.rb:11]: - [ ] This comment has not been addressed yet

File: app/models/like.rb (hunk 4 above)
Comment (descholar-ceo) [reply app/models/like.rb]: - [ ] This comment has not been addressed yet

File: app/models/comment.rb (hunk 3 above)
Comment (descholar-ceo) [reply app/models/comment.rb]: - [ ] This comment has not been addressed yet

**Review by descholar-ceo — CHANGES_REQUESTED** [review]
Hi @Ibizugbe 🤝 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
👍🏽 Communication with the reviewer is very good ✔️ 

 ## Required Changes ♻️
_Check the comments under the review._
- [ ] Kindly address the comment you got from the previous review

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Ibizugbe** [comment]
Thanks a lot @descholar-ceo  for pointing out these mistakes🙌, that was my fault😅. I made the changes on a different branch mistakenly. I have pulled from the branch now and I believe everything is OK now.

**Review by elisha2kyakpo1 — APPROVED** [review]
Hi 👋 @Ibizugbe, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights

- I like the way you communicate, keep it up💯 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Ibizugbe** [comment]
Thanks a lot @elisha2kyakpo1 for the compliments 🤗 and for taking out time to review this project🙌.

**Pull request merged**